_posts/2023-09-15-kafka-producer-client-essentials.md

```diff
@@ -0,0 +1,324 @@
+
+**`sleep` method**
+
+- Introduces a delay in the message-sending process for a specified number of milliseconds.
+- Useful for controlling the message production rate or simulating message processing times.
+- Handles potential `InterruptedException` if the thread is interrupted while paused.
+
+**`randomBytes` method**
+
+- Generates a random byte array of a specified size to serve as the payload for each message sent to the Kafka topic. Adds 65 to represent an uppercase letter in ASCII code (65 is 'A', 66 is 'B', and so on).
+- Ensures that the payload size is greater than zero and throws an `IllegalArgumentException` if not.
+
+**`retriable` method**
+
+- Determines whether to retry sending a message following a message sending error.
+- Returns `true` if the message sending process can be retried.
+- Returns `false` for null and specified exceptions, or those that do not implement the `RetriableException` interface.
+- Customizable to include other errors.
+
+**`onCompletion` method**
+
+- Confirms successful message transmission and displays information about the message sent, including the topic, partition, and offset.
+- Prints an error message on exception. Appropriate action is taken based on whether it's a fatal or non-fatal error. If the error is non-fatal, the message sending process continues. If the error is fatal, a stack trace is printed and the producer is terminated.
+
+With the imported libraries, our constants, and these configuration properties and methods, the producer client can do all we set out to do.
+
+**Example producer client**
+```java
+import java.util.Properties;
+import java.util.Random;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.apache.kafka.clients.producer.Callback;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.ProducerConfig;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.errors.RetriableException;
+import org.apache.kafka.common.serialization.ByteArraySerializer;
+import org.apache.kafka.common.serialization.LongSerializer;
+
+public class Producer implements Callback {
+    // Constants for configuration
+    private static final Random RND = new Random(0);
+    private static final String BOOTSTRAP_SERVERS = "localhost:9092";
+    private static final String TOPIC_NAME = "my-topic";
+    private static final long NUM_MESSAGES = 100;
+    private static final int MESSAGE_SIZE_BYTES = 100;
+    private static final long PROCESSING_DELAY_MS = 0L;
+
+    protected AtomicLong messageCount = new AtomicLong(0);
+
+    public static void main(String[] args) {
+        new Producer().run();
+    }
+
+    public void run() {
+        System.out.println("Running producer");
+        try (var producer = createKafkaProducer()) {
+            // Create a Kafka producer instance
```

Comment:
We can put comments above or below, but we should be consistent on that. My preference is right above.

Reply:
all on top now